### PR TITLE
Adding figure eight example and python tools package

### PR DESCRIPTION
**Pull request by cmower**
Adding a simple figure of eight example. Don't merge yet, todo:
- [x] Debug and push `figure_eight_node.py` example script
- [x] Update `lbr_examples/README.md`
- [x] Test on real robot

### CODE REVIEW

**Comment by cmower** [comment]
@mhubii this is now ready to merge.

File: lbr_description/config/fig8_config.rviz [hunk 1]
```diff
@@ -0,0 +1,209 @@
+Panels:
```
Comment (mhubii) [inline lbr_description/config/fig8_config.rviz:1]: please remove this rviz config as redundant

File: lbr_examples/README.md [hunk 2]
```diff
@@ -24,6 +24,23 @@ ros2 launch lbr_bringup lbr_bringup.launch.py model:=iiwa7 sim:=false
 ros2 run lbr_examples lbr_command_node.py
 ```
 
+### Figure of Eight
+Simple motion planning and control example. The end-effector follows a simple figure-of-eight pattern. Implementation in [scripts/figure_eight_move_to_start.py](scripts/figure_eight_move_to_start.py) and [scripts/figure_eight_node.py](scripts/figure_eight_node.py).
+- Launch robot
+```shell
+ros2 launch lbr_bringup lbr_bringup.launch.py model:=med7 sim:=true controller:=forward_position_controller rviz_config:=fig8_config
```
Comment (mhubii) [inline lbr_examples/README.md:31]: let's have a chat about this rviz config, I don't understand the need for it

File: lbr_examples/README.md [hunk 3]
```diff
@@ -24,6 +24,23 @@ ros2 launch lbr_bringup lbr_bringup.launch.py model:=iiwa7 sim:=false
...
+```
+
+- Move robot to start position
+```shell
+ros2 run lbr_examples figure_eight_move_to_start.py
```
Comment (mhubii) [inline lbr_examples/README.md:36]: can these nodes be merged? Alternatively, create launch file with conditional launch, ie on exit

File: lbr_examples/scripts/figure_eight_move_to_start.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/python3
+import sys
+import time
+import optas
```
Comment (mhubii) [inline lbr_examples/scripts/figure_eight_move_to_start.py:4]: please sort using isort, system packages should be separate

File: lbr_examples/scripts/figure_eight_move_to_start.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+import rclpy
+from lbr_python_tools.lbr_node import LBRPositionControlNode
+
+
+class FigureEightMoveToStartNode(LBRPositionControlNode):
+    def __init__(self):
+        super().__init__("figure_eight_move_to_start")
```
Comment (mhubii) [inline lbr_examples/scripts/figure_eight_move_to_start.py:11]: move node name to variabels

File: lbr_python_tools/package.xml [hunk 6]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0"?>
```
Comment (mhubii) [inline lbr_python_tools/package.xml:1]: fix this package.xml, missing license, missing description

File: lbr_examples/scripts/figure_eight_node.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
+#!/usr/bin/python3
+import time
```
Comment (mhubii) [inline lbr_examples/scripts/figure_eight_node.py:2]: sort imports

File: lbr_examples/scripts/figure_eight_node.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+        q = builder.get_model_state(self.name, -1)
+        p = self.eff_pos(q)
+        builder.add_cost_term("nominal", 10 * optas.sumsqr(q - qn))
+        builder.add_cost_term("goal_pos", 1e3 * optas.sumsqr(p - goal_pos))
+        builder.add_equality_constraint(
+            "goal_zaxis", self.zaxis(q), goal_zaxis
+        )  # goal zaxis
+        dq = builder.get_model_state(self.name, 0, time_deriv=1)
+        builder.add_cost_term("min_dq", 0.1 * optas.sumsqr(dq))
+        self.solver = optas.CasADiSolver(
```
Comment (mhubii) [inline lbr_examples/scripts/figure_eight_node.py:34]: default to consistent coding style

File: lbr_examples/scripts/figure_eight_node.py [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+            builder.build(), error_on_fail=True
+        ).setup("ipopt")
+        self.solution = None
+
+    def __call__(self, qc, goal_pos, goal_zaxis, nominal):
+        if self.solution is not None:
```
Comment (mhubii) [inline lbr_examples/scripts/figure_eight_node.py:40]: fix this double negation logic

File: lbr_examples/scripts/figure_eight_node.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+        self.solver.reset_parameters(
+            {"qc": qc, "goal_pos": goal_pos, "goal_zaxis": goal_zaxis, "qn": nominal}
+        )
+        self.solution = self.solver.solve()
+        return self.solution[f"{self.name}/q"][:, -1].toarray().flatten().tolist()
+
+
+class FigureEightNode(LBRPositionControlNode):
+    def __init__(self):
+        super().__init__("figure_eight_node")
```
Comment (mhubii) [inline lbr_examples/scripts/figure_eight_node.py:53]: move to variables

File: lbr_examples/scripts/figure_eight_node.py [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+        self.declare_parameter("hz", 100)
+        self.declare_parameter("end_effector_name", "lbr_link_ee")
+        self.declare_parameter("visualize_target", True)
+        end_effector_name = str(self.get_parameter("end_effector_name").value)
```
Comment (mhubii) [inline lbr_examples/scripts/figure_eight_node.py:57]: call get parameters values instead

File: lbr_examples/scripts/figure_eight_node.py [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+        self._goal_zaxis = None
+        self._start_time = None
+        self._controller = Controller(self.xacro_filename, dt, end_effector_name)
+        self._delay = 2.0
+
+        self._vis_pub = None
+        self._marker = Marker()
+        self._visualize_target = bool(self.get_parameter("visualize_target").value)
+        self.visualize_target = self._no_visualize
+        if self._visualize_target:
```
Comment (mhubii) [inline lbr_examples/scripts/figure_eight_node.py:70]: does this target visualization add redundancy to the other PR #66 ?

File: lbr_moveit/launch/lbr_move_group.launch.py [hunk 13]
```diff
@@ -122,8 +122,11 @@ def launch_setup(context, *args, **kwargs):
     )
 
     # RViz
+    rviz_config_filename = LaunchConfiguration("rviz_config").perform(context)
+    if not rviz_config_filename.endswith('.rviz'):
+        rviz_config_filename += '.rviz'
     rviz_config = PathJoinSubstitution(
-        [FindPackageShare("lbr_description"), "config/config.rviz"]
+        [FindPackageShare("lbr_description"), "config/" + rviz_config_filename]
```
Comment (mhubii) [inline lbr_moveit/launch/lbr_move_group.launch.py:129]: might create conflicts with the `lbr_bringup` re-write

File: lbr_python_tools/lbr_python_tools/lbr_node.py [hunk 14]
```diff
@@ -0,0 +1,158 @@
+import os
```
Comment (mhubii) [inline lbr_python_tools/lbr_python_tools/lbr_node.py:1]: could you please give me an explanation for this package so I can understand its purpose better?

**Review by mhubii — COMMENTED** [review]
Thanks for the PR, I have some questions regarding redundancy in this code base, as well as coding style and the need for `lbr_python_tools`, please elaborate

**Comment by mhubii** [comment]
I'll try to re-target this PR against one dedicated for demos, starting with humble

**Comment by mhubii** [comment]
this PR requires merge with `foxy`. 

Demos are to be moved to `advanced_python_demos`, see https://github.com/KCL-BMEIS/lbr_fri_ros2_stack/tree/humble/lbr_demos/lbr_fri_ros2_advanced_python_demos.

Will add this after #53

**Pull request closed without merging**